### PR TITLE
Typescript

### PR DESCRIPTION
**Pull request by gustawdaniel**
https://github.com/spence-s/human-object-diff/issues/16

Js -> TS

To build we need:
- ejs
- cjs
- src
- package.json

Breaking changes:

- default import replaced by named import - this is required by esm, example https://github.com/lukeed/uid/releases/tag/v2.0.0
- min node 14
- min npm 8 -> because we need "overrides" to bump "decamelize" version

### CODE REVIEW

**Comment by spence-s** [comment]
This is really great, I appreciate the refactor to TS. 

However, I want to keep this library built for node.js as it currently is. 

Please remove esbuild, and use tsc to create the build. Target Node16 as your output and emit d.ts files as well so type defs are included. Add anything to support this to package.json that is needed. 

I'm fine with moving tests to jest, but please just run it with a tsx loader.

**Review by spence-s — CHANGES_REQUESTED** [review]
Remove esbuild - use tsc to target commonjs and Node16.

**Comment by spence-s** [comment]
Also, lets keep the default export - there is no reason to change that.

File: README.md [hunk 1]
```diff
@@ -51,20 +53,37 @@ console.log(diff(lhs, rhs));
 // -> ['"Foo", with a value of "bar" (at Obj.foo) was changed to "baz"']
 ```
 
+ES Module
+
 ## Configuring
 
+```typescript
+(async () => {
```
Comment (spence-s) [inline README.md:61]: this example provides no value. You can do this pattern in cjs or esm and there is no reason to show that. 

cjs
```js
const HR = require('human-object-diff')
```
esm
```js
import HR from 'human-object-diff'
```
dynamic import in esm
```js
const {default: HR} = await import("human-object-diff")
```
dynamic import in cjs
```js
(async() =>{
  const {default: HR} = await import("human-object-diff")
})()
```

All of the above is supported by just exporting a single cjs build.

**Comment by gustawdaniel** [comment]
@spence-s I will remove esm as you wish.

But please explain me, what advantage has tsc over esbuild?

For me esbuild is better because is few times faster. I am really curious about reason, why somebody can prefer tsc.

**Comment by gustawdaniel** [comment]
@spence-s I will remove esm as you wish.

But please explain me, what advantage has tsc over esbuild?

For me esbuild is better because is few times faster. I am really curious about reason, why somebody can prefer tsc.

**Comment by gustawdaniel** [comment]
I stuck with the following problem:

In index.ts i can write

```
class DiffEngine {}
module.exports = DiffEngine;
```

and then:

```
const HumanDiff = require('human-object-diff')
```

works but:

```
import HumanDiff from 'human-object-diff';
```
not

```
import * as  HumanDiff from 'human-object-diff';
```

or

```
const  HumanDiff = await import('human-object-diff');
```

is

```
console.log(HumanDiff);
```

and i have to write ts-ignore to use it

```
// @ts-ignore
const { diff } = new HumanDiff.default(options);
```

---

but on other hand when I write

```
```

then after

```
const HumanDiff = require('human-object-diff')
```

i have to break compatibility and use as

```
const { diff } = new HumanDiff.default(options);
```
 but for
 
```
import HumanDiff from 'human-object-diff'; 
const  HumanDiff = await import('human-object-diff');
```

it works as before

**Comment by gustawdaniel** [comment]
Fixed by 

```
// CommonJS export
module.exports = DiffEngine;
module.exports.default = DiffEngine;

// ES module export
export default DiffEngine;
```

**Comment by spence-s** [comment]
You only need to do:  `export default DiffEngine` in ts. 

You do not need to use module.exports at all or module.exports.default

Note: CJS is a compile target from typescript, you still use import/export in the ts code.

**Pull request merged**